grafana/grafana
# Elasticsearch: Fix processing of logs with not-recognized time format

This is a bug fix for processing time field in ES on backend, when time field uses format that golang does not recognize (e.g. MM/DD/YYY, unix timestamp with nanoseconds decimal point, ...). To fix this, for log queries, we are using `fields` to get time field in `strict_date_optional_time_nanos` format that is recognizable by golang and also supports nanosecond precision.

This is only affecting logs queries.

I have added and updates tests for this. Moreover I have added `@timestamp_custom`, `@timestamp_unix` and `@timestamp_nanos` to `data.js` for easier testing of this PR, but also for easier testing of time formats in the future.

Fixes https://github.com/grafana/grafana/issues/67391

To test this:

1. Run `make devenv sources=elastic`
2. Make sure that `enableElasticsearchBackendQuerying` is set to `true`
3. Create a new elastic data source pointing to `https://localhost:9200` and for time field add `@timestamp_custom`. 
4. Run log query - you should not see `Invalid date` in time column
5. You can also switch to `@timestamp`, `@timestamp_unix`, `@timestamp_nanos` to ensure all formats work as expected.

Fixed: 

<img width="1476" alt="image" src="https://user-images.githubusercontent.com/30407135/235988129-fcdbf9b1-7478-45b7-9001-b951fab8515c.png">

Current main (broken):

<img width="1474" alt="image" src="https://user-images.githubusercontent.com/30407135/235988479-66beb0d1-cc30-433e-8332-44f74a049c53.png">
